### ISSUE TITLE
Optional return type masks incompatible use of Any

### ISSUE THREAD

**Issue by GPHemsley** [python/mypy#9213]
Given this script:
```python
#!/usr/bin/env python3

from typing import (
	Any,
	Dict,
	Optional,
)

def required_int() -> int:
	x: Dict[str, Any] = {
		"id": "12345",
	}

	return x.get("id")

def optional_int() -> Optional[int]:
	x: Dict[str, Any] = {
		"id": "12345",
	}

	return x.get("id")
```

Running `mypy --strict` gives the following error for `required_int`:
```
14: error: Incompatible return value type (got "Optional[Any]", expected "int")
```

However, it does not throw any error for `optional_int`! I would expect to get the same error for both.

(Note: Using `Union[int, None]` has the same effect, so this isn't specific to `Optional`.)

Tested on:
Python 3.6.11
mypy 0.782

**Comment by GPHemsley** [comment]
Here's another example where the interaction with `None` seems to absolve the `Any` type mismatch:
```python
#!/usr/bin/env python3

from typing import (
	Any,
	Dict,
	Optional,
)

def required_int(x: Dict[str, Any]) -> int:
	x_id = x.get("id")
	# x_id: Optional[object] = x.get("id")
	# x_id: Any = x.get("id")
	# x_id = x["id"]

	return int(x_id)

def optional_int(x: Dict[str, Any]) -> Optional[int]:
	x_id = x.get("id")
	# x_id: Optional[object] = x.get("id")
	# x_id: Optional[Any] = x.get("id")
	# x_id = x["id"]

	if x_id is None:
		return None

	return int(x_id)
```

`required_int` throws an error, but `optional_int` does not:
```
15: error: Argument 1 to "int" has incompatible type "Optional[Any]"; expected "Union[str, bytes, SupportsInt, _SupportsIndex]"
```

**Comment by Akuli** [comment]
mypy is correct here because `x.get("key that does not exist")` returns None rather than raising an exception,  and None is not a valid int

maybe you want `x["id"]` instead of `x.get("id")`?

**Comment by GPHemsley** [comment]
@Akuli The problem is not with `x.get("key that does not exist")`, it's with `x.get("key that DOES exist")`.

**Comment by Akuli** [comment]
How is mypy supposed to know which keys exist in the dict if all you tell is that it's `Dict[str, something]`?

`TypedDict` is a dict whose keys are strings known to mypy.

**Comment by Akuli** [comment]
I think I finally understood your first example. Sorry for being dumb. `required_int()` errors because `x.get("id")` returns `Optional[Any]`, aka `Union[None, Any]`. Generally, returning a union from a function means that every union member must be a valid return value. In this case, `None` must be a valid `int` (it isn't so you get the error) and `Any` must be a valid `int` (it is). But both `None` and `Any` are valid `Optional[Any]` objects, so mypy doesn't see that there's something wrong if your function returns `Optional[Any]`.

If you return `x["id"]` instead, then neither function will cause errors, because `x["id"]` has type `Any`, and `Any` values are valid `int`s and valid `Optional[int]`s. So, returning a value with type `Any` and returning a value with type `Union[None, Any]` are not the same, because of how unions work.

Looking at your second example code, I don't know why you would expect its `optional_int` function to error. By checking `if x_id is None`, you change the type of `x_id` from `Union[None, Any]` to `Any`, and then it's fine to pass it to `int` because mypy knows that you won't end up doing `int(None)`. Why should the `optional_int` function of your second example create a mypy error?

**Comment by GPHemsley** [comment]
@Akuli Among other things, `int(object())` is incompatible with the type signature of `int()`, so "`Any` except `None`" is not enough to be valid.

**Comment by Akuli** [comment]
This code gives a mypy error as expected:

```
def optional_int(x: Dict[str, Any]) -> Optional[int]:
    x_id: Optional[object] = x.get("id")
    if x_id is None:
        return None
    return int(x_id)
```

This is how `Any` and `object` differ. `Any` means that mypy won't complain about how you use the value, and `object` means that you are only allowed to do things that work with all objects. You can't `int()` some objects, so mypy errors.

What exactly do you mean with "`Any` except `None`"? `Any` means that you can do anything with the value, and `Optional[Any]` means that you can do anything with the value as long as you check for `None`-ness.

**Comment by GPHemsley** [comment]
> `Any` means that mypy won't complain about how you use the value, and `object` means that you are only allowed to do things that work with all objects. You can't `int()` some objects, so mypy errors.

> `Any` means that you can do anything with the value, and `Optional[Any]` means that you can do anything with the value as long as you check for `None`-ness.

If I am understanding correctly what you are saying, this is not quite how I have internalized the meaning of `Any`.

To my mind, `Any` means "any possible type"; by defining `x` as `Dict[str, Any]`, I am saying that the values in `x` could be any possible type – `int`, `str`, `object`, `None`, `Whatever`. It is my understanding that `Any` and `Optional[Any]` are the same thing. If I explicitly check for `None` in a value with that type, then all I'm doing is narrowing `Any` (or `Optional[Any]`) to "`Any` except `None`" – it can still be `int`, `str`, `object`, or `Whatever`. With a return type of `Optional[int]`, most of those types are still incompatible, and so an error should still be thrown.

Are you saying that I am misunderstanding how things are intended to work? Do you disagree that there is a bug here?

**Comment by Akuli** [comment]
`Any` can be thought of as "any possible type", but that's too vague for understanding what mypy is doing in this case.

You should instead think of `Any` as "mypy, please don't complain". `Optional[Any]` then means "mypy, please don't complain unless I forget to check for None".

"Narrowing down" `Any` doesn't change it, and mypy won't complain about how you use it, even after narrowing down. After all, that's what you asked mypy to do by using `Any`. Consider this, for example:

```python3
def bar() -> Any: ...

def foo() -> None:
    x = bar()
    reveal_type(x)  # Any
    assert not isinstance(x, int)
    assert x is not None
    reveal_type(x)  # still Any
```

You can still narrow down like this though:

```python3
    x = bar()
    reveal_type(x)  # Any
    assert isinstance(x, int)
    reveal_type(x)  # now it's int
```

**Comment by Akuli** [comment]
tl;dr: I still haven't seen anything that would look like a bug to me. While "`Any` means any type" is correct, you also need a more precise "`Any` means mypy shuts up and doesn't complain" understanding for debugging `Any` related problems.

**Comment by GPHemsley** [comment]
Allow me to present my real-world scenario, then:
* I have been unknowingly passing around a `str` in all sorts of places defined as requiring `Optional[int]` because it originated as a value in `Dict[str, Any]`. My first example is a reduced testcase for that.
* My attempt to work around this without giving up `Dict[str, Any]`, and to force mypy to throw an error in the correct place, led to my second example. (I forget exactly what undesirable behavior my real-world code was causing, but suffice to say it was enough to leave me unsatisfied that it would sufficiently address my issue.)

If that is not a bug, what do you propose I do to cause mypy to yell when it tries to return a value with a not-guaranteed-to-be-`Optional[int]` from a function with a return type of `Optional[int]`, and also when such value is passed to another function as a parameter that is also of type `Optional[int]`?

**Comment by Akuli** [comment]
How does data arrive into your program? What's the first occurence of `Any`?

**Comment by Akuli** [comment]
Hm. If you want to use `Dict[str, Any]`, and the values are often of type `str` (but not always in which case you actually need that `Any`), you should do something like this:

```python3
def i_want_string(foo: str) -> None: ...

i_want_string(str_any_dict['some key'])
```

mypy won't complain.

Use `.get()` only if the key may be missing, and then check for the `None` that it might return (mypy will warn if you forget). You can also provide a fallback:

```python3
i_want_string(str_any_dict.get('some key', 'this will be used if key not found'))
```

**Comment by gvanrossum** [comment]
The problem can be reduced to the following:
```
a: Any = "xyz"
b: int = a  # No error from mypy
```
This is how `Any` works.

**Comment by Akuli** [comment]
Maybe I should also mention the way I work with Any types: get rid of the `Any` type as soon as possible, and the rest of your code won't be confused. For example, you can use `assert isinstance(some_variable, int)` like I showed earlier.

If you want to require `assert isinstance(some_variable, the_type_you_expect)` whenever that dict is used, you can make it `Dict[str, object]` instead of `Dict[str, Any]`.

**Comment by GPHemsley** [comment]
I wasn't aware that using `assert isinstance` was considered a best practice for dealing with types; that might be enough to solve my real-world problem.

However, I'm still not clear how my initial example is not a bug.

----

Consider the following:
```python
#!/usr/bin/env python3

from typing import (
	Any,
	Dict,
	Union,
)


def required_int() -> int:
	x: Dict[str, Any] = {
		"id": "12345",
	}

	y = x.get("id")  # Union[Any, None]

	return y  # Union[Any, None] != int


def optional_int() -> Union[int, None]:
	x: Dict[str, Any] = {
		"id": "12345",
	}

	y = x.get("id")  # Union[Any, None]

	return y  # Union[Any, None] != Union[int, None]
```

This reports:
> 17: error: Incompatible return value type (got "Optional[Any]", expected "int")

It does _not_ report:
> 27: error: Incompatible return value type (got "Optional[Any]", expected "Optional[int]")

----

But then:
```python
#!/usr/bin/env python3

from typing import (
	Any,
	Dict,
	Union,
)


def required_int() -> int:
	x: Dict[str, Any] = {
		"id": "12345",
	}

	y = x.get("id")  # Union[Any, None]

	assert y is not None  # Any

	return y  # Any != int


def optional_int() -> Union[int, None]:
	x: Dict[str, Any] = {
		"id": "12345",
	}

	y = x.get("id")  # Union[Any, None]

	assert y is not None  # Any

	return y  # Any != Union[int, None]
```

This reports:
> 19: error: Returning Any from function declared to return "int"
> 31: error: Returning Any from function declared to return "Optional[int]"

----

So it knows that `Any` is unacceptable in 3 of the 4 cases. Why does it not know about the 4th case?

I'm not explicitly telling it that `y` is of type `Union[int, None]`, so this doesn't seem to me like a case of compatible redefinition of the variable type.

**Comment by Akuli** [comment]
Running mypy with `--strict --show-error-codes`, I noticed that the functions in your second example code create errors because of an error code named `[no-any-return]`. To me it seems to literally disallow returning a value of type `Any`, and it doesn't detect returning `Optional[Any]`.

Maybe it is a bug after all?

**Comment by gvanrossum** [comment]
Oops, I missed that the issue is about the behavior of `--strict`. A shorter repro is (with `mypy --strict`):
```
from typing import *
def f(a: Any) -> int:
    return a  # error: Returning Any from function declared to return "int"  [no-any-return]
def g(a: Optional[Any]) -> Optional[int]:
    return a  # No error
```
I guess that's a legitimate issue with the code responsible for `[no-any-return]`.

**Comment by GPHemsley** [comment]
So something in here then?

https://github.com/python/mypy/blob/1bbcd536a59ce07fad78f6ec12b53e3859a03aaf/mypy/checker.py#L3194-L3206

**Comment by gvanrossum** [comment]
Well, it'll end up taking the branch where it calls `self.check_subtype()` and that's quite the can of worms.

**Comment by JukkaL** [comment]
I guess we could complain if there is an `Any` type component anywhere in the type of the returned value.

**Comment by Akuli** [comment]
It doesn't make sense to complain when the return type is a class with `Any` as the type of one of the members. Maybe "anywhere in the type of the returned type" should check for unions but not much more?

**Comment by hauntsaninja** [comment]
Closing, since mypy's behaviour is clearly correct in the original examples. I think changes to `--warn-return-any` are likely to be disruptive, so I'd like to see more evidence that checking `Any` is a useful feature request before making changes.
